mmengine/analysis/jit_analysis.py

```diff
@@ -0,0 +1,672 @@
+    'aten::flatten',
+    'aten::floor',
+    'aten::floor_divide',
+    'aten::full',
+    'aten::full_like',
+    'aten::gather',
+    'aten::ge',
+    'aten::gt',
+    'aten::index',
+    'aten::index_put_',
+    'aten::masked_fill',
+    'aten::max',
+    'aten::narrow',
+    'aten::new_empty',
+    'aten::new_full',
+    'aten::new_zeros',
+    'aten::nonzero',
+    'aten::ones',
+    'aten::permute',
+    'aten::relu',
+    'aten::relu_',
+    'aten::remainder',
+    'aten::reshape',
+    'aten::roll',
+    'aten::select',
+    'aten::size',
+    'aten::slice',
+    'aten::split',
+    'aten::split_with_sizes',
+    'aten::squeeze',
+    'aten::stack',
+    'aten::t',
+    'aten::to',
+    'aten::transpose',
+    'aten::type_as',
+    'aten::unbind',
+    'aten::unsqueeze',
+    'aten::unsqueeze_',
+    'aten::view',
+    'aten::zeros',
+    'aten::zeros_like',
+}
+
+
+@dataclass
+class Statistics:
+    """For keeping track of the various model statistics recorded during
+    analysis."""
+
+    counts: 'Dict[str, typing.Counter[str]]'
+    unsupported_ops: 'Dict[str, typing.Counter[str]]'
+    uncalled_mods: 'Set[str]'
+
+
+def _named_modules_with_dup(model: nn.Module,
+                            prefix: str = ''
+                            ) -> Iterable[Tuple[str, nn.Module]]:
+    """The same as `model.named_modules()`, except that it includes duplicated
+    modules that have more than one name."""
+    yield prefix, model
+    for name, module in model._modules.items():
+        if module is None:
+            continue
+        submodule_prefix = prefix + ('.' if prefix else '') + name
+        yield from _named_modules_with_dup(module, submodule_prefix)
+
+
+def _named_modules_without_dup(
+        model: nn.Module) -> Iterator[Tuple[str, nn.Module]]:
+    """Like .named_modules(), but the results are slightly different for some
+    wrapped models."""
+    seen = set()
+    for name, mod in _named_modules_with_dup(model):
+        if mod not in seen:
+            seen.add(mod)
+            yield name, mod
+
+
+def _get_scoped_trace_graph(
+    module: nn.Module,
+    inputs: Union[Tensor, Tuple[Tensor, ...]],
+    aliases: Dict[Union[str, nn.Module], str],
+) -> torch._C.Graph:
+    """Traces the provided module using torch.jit._get_trace_graph, but adds
+    submodule scope information to each graph node.
+
+    The resulting graph
+    is in-lined and has all model parameters treated as inputs. The input
+    model has the scope name '', while its descendants have names of the
+    form 'child.grandchild.grandgrandchild...'.
+
+    Args:
+        model (nn.Module) : The module to trace
+        inputs (tuple) : Inputs used during the trace of the model
+        aliases (dict(str or nn.Module, str) : maps modules and module
+            names to the canonical name to be used as the scope for
+            that module.
+
+    Returns:
+        graph (torch._C.Graph) : The pytorch JIT trace of the model
+    """
+
```

Comment:
Add some comments:

# torch.jit._get_trace_graph can trace torch function like `aten::linear`, `aten::add` etc. However, the traced node(function) cannot tell it is called by which module. `ScopePushHook` and `ScopePopHook` can help traced node get the module name information by `node.scopeName()`.